src/Task/TalkNPC.pm

```diff
@@ -206,103 +298,137 @@ sub iterate {
 			}
-
-		} elsif ( $step =~ /^r(?:(\d+)|=(.+)|~\/(.*?)\/(i?))/i ) {
-			# Choose a menu item.
-			my $choice = $1;
-			if ($npcTalkType eq 'select' and $2 || $3) {
-				# Choose a menu item by matching options against a regular expression.
-				my $pattern = $2 ? "^\Q$2\E\$" : $3;
-				my $postCondition = $4;
-				( $choice ) = grep { $postCondition ? $talk{responses}[$_] =~ /$pattern/i : $talk{responses}[$_] =~ /$pattern/ } 0..$#{$talk{responses}};
-				if (defined $choice && $choice < $#{$talk{responses}}) {
-					$messageSender->sendTalkResponse($talk{ID}, $choice + 1);
-				} elsif (defined $choice) {
-					# The last response is a fake "Cancel Chat" response.
-					$self->cancelTalk();
-				} else {
-					$self->setError(WRONG_NPC_INSTRUCTIONS,
-						TF("According to the given NPC instructions, a menu " .
-						"item matching '%s' must now be selected, but no " .
-						"such menu item exists.", $pattern));
-					$self->cancelTalk();
-				}
-			} elsif ($npcTalkType eq 'select') {
-				if ($choice < $#{$talk{responses}}) {
-					$messageSender->sendTalkResponse($talk{ID}, $choice + 1);
-				} elsif ($choice) {
-					# The last response is a fake "Cancel Chat" response.
-					$self->cancelTalk();
-				} else {
-					$self->setError(WRONG_NPC_INSTRUCTIONS,
-						TF("According to the given NPC instructions, menu item %d must " .
-						"now be selected, but there are only %d menu items.",
-						$choice, @{$talk{responses}} - 1));
-					$self->cancelTalk();
-				}
+			
+		# Send NPC talk text.
+		} elsif ($current_talk_step eq 'text') {
+			if ( $step =~ /^t=(.*)/i ) {
+				$messageSender->sendTalkText($talk{ID}, $1);
+			
+			# Wrong sequence
 			} else {
-				$self->setError(WRONG_NPC_INSTRUCTIONS,
-					T("According to the given NPC instructions, a menu item " .
-					"must now be selected, but that's not possible."));
-				$self->cancelTalk();
+				$self->set_error_and_try_to_cancel(WRONG_NPC_INSTRUCTIONS, T("This npc requires a text to be sent now, but the given instructions don't match that."));
 			}
-
-		} elsif ( $step =~ /^n/i ) {
-			# Click Close or Cancel.
-			$self->cancelTalk();
-			$ai_v{npc_talk}{time} = time;
-			$self->{time} = time;
-
-		} elsif ( $step =~ /^b.*/i ) {
-			# Get the shop's item list.
-			if ($step =~ /^b$/i) {
+		
+		# Get the sell or buy list in a shop.
+		} elsif ( $current_talk_step eq 'buy_or_sell' ) {
+		
+			# Get the sell list in a shop.
+			if ( $step =~ /^s/i ) {
+				$messageSender->sendNPCBuySellList($talk{ID}, 1);
+				
+			# Get the buy list in a shop.
+			} elsif ($step =~ /^b$/i) {
 				$messageSender->sendNPCBuySellList($talk{ID}, 0);
-			# Bulk buy solution.
-			} elsif ($step =~ /^b(\d+),(\d+)/i) {
+				
+			# Click the cancel button in a shop.
+			} elsif ($step =~ /^e$/i) {
+				$ai_v{'npc_talk'}{'talk'} = 'close';
+				$self->conversation_end;
```

Comment:
"e" shouldn't click the cancel button and it doesn't seem it's done there; so, wrong comment?